addons/mail/models/mail_thread.py

```diff
@@ -835,6 +836,75 @@ def _routing_reset_bounce(self, email_message, message_dict):
             for model in [bl_model for bl_model in bl_models if bl_model.model in self.env]:  # transient test mode
                 self.env[model.model].sudo().search([('message_bounce', '>', 0), ('email_normalized', '=', valid_email)])._message_reset_bounce(valid_email)
 
+    @api.model
+    def _routing_detect_loop_model(self, message_dict, routes):
+        """This method returns True is the incoming email should be ignored.
+
+        The goal of this method is to prevent loops which can occur if an auto-replier
+        send email to Odoo.
+        """
+        # Detect the email address sent to many emails
+        get_param = self.env['ir.config_parameter'].sudo().get_param
+        INCOMING_LIMIT_PERIOD = int(get_param('mail.incoming.limit.period', 60))
+        INCOMING_LIMIT_ALIAS = int(get_param('mail.incoming.limit.alias', 5))
+
+        create_date_limit = datetime.datetime.now() - datetime.timedelta(minutes=INCOMING_LIMIT_PERIOD)
+
+        for model, thread_id, *__ in routes or []:
+            # Check if we create a record by an alias
+            if thread_id:
+                # Reply to an existing thread
+                continue
+
+            email_from = message_dict.get('email_from')
+            if not email_from:
+                continue
+
+            model = self.env[model]
+
+            if 'mail.thread.blacklist' in model._inherit:
+                domain = [('email_normalized', '=', tools.email_normalize(email_from))]
+            else:
+                primary_email = getattr(model, '_primary_email', None)
+                if not primary_email or primary_email not in model._fields:
+                    _logger.info('Primary email missing on %s', model._name)
+                    continue
+
+                domain = [(primary_email, 'ilike', tools.email_normalize(email_from))]
+
+            mail_incoming_messages_count = model.sudo().search_count(
+                expression.AND([
+                    [('create_date', '>', create_date_limit)],
+                    domain,
+                ]),
+            )
+
+            if mail_incoming_messages_count >= INCOMING_LIMIT_ALIAS:
+                _logger.info('This email address created too many <%s>.', model)
+                return True
+
+        return False
+
+    @api.model
+    def _routing_detect_loop(self, message):
```

Comment:
``_routing_detect_loop_from_headers`` ? That way we easily understand its purpose and why it is not model / route based.